### ISSUE TITLE
[Inductor] [CPU] Low efficient transpose in shufflenet_v2_x1_0

### ISSUE THREAD

**Issue by Valentine233** [pytorch/torchdynamo#1915]
### Description
Comparing performances of `shufflenet_v2_x1_0` with backends inductor and IPEX, inductor is 0.74 IPEX. One main reason is the low efficient transpose, the other one is reported in https://github.com/pytorch/torchdynamo/issues/1914.

### Profiling and Code snippet
`kernel_cpp_4` is for `torch.cat + torch.transpose.contiguous` implementation.
![image](https://user-images.githubusercontent.com/23010269/203484815-8bcab50f-f33f-4246-b097-d15e4eff12ce.png)
```
kernel_cpp_4 = async_compile.cpp('''
#include "/home/neo/workspace/workspace01_dynamotest/pytorch/torch/_inductor/codegen/cpp_prefix.h"
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       const float* __restrict__ in_ptr1,
                       const float* __restrict__ in_ptr2,
                       float* __restrict__ out_ptr0,
                       float* __restrict__ out_ptr1,
                       float* __restrict__ out_ptr2,
                       float* __restrict__ out_ptr3)
{
    RECORD_FUNCTION(__FUNCTION__, c10::ArrayRef<c10::IValue>({})); 
    #pragma omp parallel num_threads(28)
    {
        #pragma omp for 
        for(long i0=0; i0<64; i0+=1)
        {
            #pragma GCC ivdep
            for(long i1=0; i1<58; i1+=1)
            {
                for(long i2=0; i2<49; i2+=1)
                {
                    auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + (16*i2) + (784*(i1 / 2)) + (45472*(i1 % 2)) + (90944*i0));
                    tmp0.store(out_ptr0 + (16*i2) + (784*i1) + (90944*i0));
                }
                #pragma omp simd simdlen(8) 
                for(long i2=784; i2<784; i2+=1)
                {
                    auto tmp0 = in_ptr0[i2 + (784*(i1 / 2)) + (45472*(i1 % 2)) + (90944*i0)];
                    out_ptr0[i2 + (784*i1) + (90944*i0)] = tmp0;
                }
            }
        }
        #pragma omp for 
        for(long i0=0; i0<64; i0+=1)
        {
            #pragma GCC ivdep
            for(long i1=0; i1<58; i1+=1)
            {
                #pragma GCC ivdep
                for(long i2=0; i2<784; i2+=1)
                {
                    {
                        {
                            auto tmp0 = in_ptr1[i1 + (58*i2) + (45472*i0)];
                            out_ptr1[i2 + (784*i1) + (90944*i0)] = tmp0;
                        }
                    }
                }
            }
        }
        #pragma omp for 
        for(long i0=0; i0<64; i0+=1)
        {
            #pragma GCC ivdep
            for(long i1=0; i1<116; i1+=1)
            {
                #pragma GCC ivdep
                for(long i2=0; i2<784; i2+=1)
                {
                    {
                        {
                            auto tmp0 = in_ptr2[i2 + (784*(i1 / 2)) + (45472*(i1 % 2)) + (90944*i0)];
                            out_ptr2[i1 + (116*i2) + (90944*i0)] = tmp0;
                            out_ptr3[i1 + (116*i2) + (90944*i0)] = tmp0;
                        }
                    }
                }
            }
        }
    }
}
''')
```

### Minified repro
`python benchmarks/dynamo/torchbench.py --performance --float32 -dcpu -n50 --inductor --no-skip --dashboard -k "shufflenet_v2_x1_0" --cold_start_latency --channels-last`

cc @EikanWang @jgong5

**Comment by jgong5** [comment]
Most of the generated kernels have transposition operations except for kernel_1 which requires vectorized masked load support. I grouped these kernels except for kernel_1 with the profiles below. These kernels take larger portion compared to kernel_1 with multi-threaded inference than single-threaded inference.

Single thread single batch:
![image](https://user-images.githubusercontent.com/8359223/206906164-33f65f2f-e100-4358-badb-11a8ff0d2ee4.png)

Multi-thread bs=64:
![image](https://user-images.githubusercontent.com/8359223/206906140-9f93a7d6-8eff-4820-8718-721fd58e4925.png)

**Comment by jgong5** [comment]
A quick [POC](https://github.com/jgong5/pytorch/tree/jgong5/inductor_transpose) brought the performance result below (tested on ICX).

## ChannelShuffle op benchmark

```python
    def forward(self, arg0_1):
        view = torch.ops.aten.view.default(arg0_1, [64, 2, 58, 28, 28]);  arg0_1 = None
        permute = torch.ops.aten.permute.default(view, [0, 2, 1, 3, 4]);  view = None
        clone = torch.ops.aten.clone.default(permute, memory_format = torch.contiguous_format);  permute = None
        view_1 = torch.ops.aten.view.default(clone, [64, -1, 28, 28]);  clone = None
        return (view_1,)
```
Input: batch_size: 64, channels_per_group: 58, height: 28, width: 28, groups: 2, channel_last: True

ChannelShuffle UT | Eager | Inductor (before) | Inductor (after) | Speedup (before vs. eager) | Speedup (after vs. before)
-- | -- | -- | -- | -- | --
Single-thread (us) | 169.8 | 173.2 | 135.9 | 0.98 | 1.27
Multi-thread (32c) (us) | 14408.5 | 14021.8 | 11566.6 | 1.03 | 1.21

## shufflenet_v2_x1_0

shufflenet_v2_x1_0 | Eager | Inductor (before) | Inductor (after) | Speedup (before vs. eager) | Speedup (after vs. before)
-- | -- | -- | -- | -- | --
Single-thread (bs=1) (ms) | 26.3 | 13.8 | 13.3 | 1.91 | 1.04
Multi-thread (32c, bs=64) (ms) | 38.5 | 19.6 | 18.8 | 1.96 | 1.04

Note that the transpose load/store originally accounts for 8% of `shufflenet_v2_x1_0`. The optimization brings 2X to that part.

## Generated code

Some generated loops from ChannelShuffle. `TILE2D_STORE` leverages `transpose_kernel_16x16_avx512` kernel in FBGEMM. `TILE2D_LOAD` is a direct vector load implementation.

Before optimization
```c++
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       float* __restrict__ out_ptr0)
{
    #pragma GCC ivdep
    for(long i0=0; i0<64; i0+=1)
    {
        #pragma GCC ivdep
        for(long i1=0; i1<58; i1+=1)
        {
            #pragma GCC ivdep
            for(long i2=0; i2<2; i2+=1)
            {
                #pragma GCC ivdep
                for(long i3=0; i3<784; i3+=1)
                {
                    {
                        {
                            auto tmp0 = in_ptr0[i1 + (58*i2) + (116*i3) + (90944*i0)];
                            out_ptr0[i3 + (784*i2) + (1568*i1) + (90944*i0)] = tmp0;
                        }
                    }
                }
            }
        }
    }
}
```

After optimization

```c++
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       float* __restrict__ out_ptr0)
{
    #pragma omp parallel num_threads(32)
    {
        #pragma omp for 
        for(long i0=0; i0<64; i0+=1)
        {
            #pragma GCC ivdep
            for(long i1=0; i1<3; i1+=1)
            {
                #pragma GCC ivdep
                for(long i2=0; i2<2; i2+=1)
                {
                    #pragma GCC ivdep
                    for(long i3=0; i3<49; i3+=1)
                    {
                        {
                            float tmp0[16*16] __attribute__ ((aligned (16))); TILE2D_LOAD(tmp0, in_ptr0 + (16*i1) + (58*i2) + (116*i1_inner) + (1856*i3) + (90944*i0), i1_inner, 16, float);
                            TILE2D_STORE(out_ptr0 + (16*i3) + (784*i2) + (25088*i1) + (90944*i0), tmp0, 1568, 16);
                        }
                    }
                    #pragma GCC ivdep
                    for(long i3=784; i3<784; i3+=1)
                    {
                        for (long i1_inner = 0; i1_inner < 16; i1_inner++)
                        {
                            auto tmp0 = in_ptr0[i1_inner + (16*i1) + (58*i2) + (116*i3) + (90944*i0)];
                            out_ptr0[i3 + (784*i2) + (1568*i1_inner) + (25088*i1) + (90944*i0)] = tmp0;
                        }
                    }
                }
            }
            #pragma GCC ivdep
            for(long i1=48; i1<58; i1+=1)
            {
                #pragma GCC ivdep
                for(long i2=0; i2<2; i2+=1)
                {
                    #pragma GCC ivdep
                    for(long i3=0; i3<784; i3+=1)
                    {
                        {
                            auto tmp0 = in_ptr0[i1 + (58*i2) + (116*i3) + (90944*i0)];
                            out_ptr0[i3 + (784*i2) + (1568*i1) + (90944*i0)] = tmp0;
                        }
                    }
                }
            }
        }
    }
}
```
